examples/setixset.dx

```diff
@@ -0,0 +1,243 @@
+  s = size n
+  case s == 0 of
+    True -> Nothing
+    False -> Just xs.(unsafeFromOrdinal n 0)
+
+def allExceptLast {n a} (xs:n=>a) : List a =
+  shortSize = Fin (max 0 ((size n) - 1))
+  allButLast = view i:shortSize. xs.(unsafeFromOrdinal _ (ordinal i))
+  (AsList _ allButLast)
+
+def mergeUniqueSortedLists {a} [Eq a] (xlist:List a) (ylist:List a) : List a =
+    -- This function is associative, for use in a monoidal reduction.
+    -- Assumes all xs are <= all ys.
+    -- The element at the end of xs might equal the
+    -- element at the beginning of ys.  If so, this
+    -- function removes the duplicate when concatenating the lists.
+    (AsList nx xs) = xlist
+    (AsList _  ys) = ylist
+    case last xs of
+      Nothing -> ylist
+      Just last_x -> case first ys of
+        Nothing -> xlist
+        Just first_y -> case last_x == first_y of
+          False -> concat [xlist,            ylist]
+          True ->  concat [allExceptLast xs, ylist]
+
+def removeDuplicatesFromSorted {n a} [Eq a] (xs:n=>a) : List a =
+  xlists = for i:n. (AsList 1 [xs.i])
+  reduce (AsList 0 []) mergeUniqueSortedLists xlists
+
+
+'#### Misc instances
+
+instance {n} [Ix n] Eq n
+  (==) = \x y. ordinal x == ordinal y
+
+def lexicalOrder {n} [Ord n]
+  (compareElements:n->n->Bool)
+  (compareLengths:Int->Int->Bool)
+  ((AsList nx xs):List n) ((AsList ny ys):List n) : Bool =
+    -- Orders Lists according to the order of their elements,
+    -- in the same way a dictionary does.
+    -- For example, this lets us sort Strings.
+    --
+    -- This function operates serially and short-circuits
+    -- at the first difference.  One could also write this
+    -- functon to operate in parallel, but it would be
+    -- wasteful in the case where there is an early difference.
+    iter \i.
+      case i == min nx ny of
+        True -> Done $ compareLengths nx ny
+        False ->
+          xi = xs.(unsafeFromOrdinal _ i)
+          yi = ys.(unsafeFromOrdinal _ i)
+          case compareElements xi yi of
+            True -> Done True
+            False -> case xi == yi of
+              True -> Continue
+              False -> Done False
+
+instance {n} [Ord n] Ord (List n)
+  (>) = lexicalOrder (>) (>)
+  (<) = lexicalOrder (<) (<)
+
+
+
+'### Sets
+
+data Set a [Ord a] =
+  -- Guaranteed to be in sorted order with unique elements,
+  -- as long as no one else uses this constructor.
+  -- Instead use the "toSet" function below.
+  UnsafeAsSet n:Int elements:(Fin n => a)
+
+def toSet {n a} [Ord a] (xs:n=>a) : Set a =
+  sorted_xs = sort xs
+  (AsList n' sorted_unique_xs) = removeDuplicatesFromSorted sorted_xs
+  UnsafeAsSet _ $ unsafeCastTable (Fin n') sorted_unique_xs
+
+def setSize {a} ((UnsafeAsSet n _):Set a) : Int = n
+
+instance {a} [Eq a] Eq (Set a)
+  (==) = \(UnsafeAsSet nx xs) (UnsafeAsSet ny ys).
+    if nx /= ny
+      then False
+      else all for i:(Fin nx).
+        xs.i == ys.(unsafeFromOrdinal _ (ordinal i))
```

Comment:
A neat way to implement this (no need to change it for now) would be to write a function `equalSizeFin : n:Int -> m:Int -> Maybe (Iso (Fin n) (Fin m))` that returns the isomorphism if `n == m`. The isomorphism would contain the `unsafeFromOrdinal` inside it, but any use of this code would be entirely safe!

Reply:
That'd be cool.  In the meantime, I realized I could just cast the sets to list and re-use list equality.  But the list equality code in the prelude could use this (although it'd need to be moved to after `Iso` is defined).